### PR TITLE
#1569 implemetation

### PR DESCRIPTION
**Pull request by adimova**
Signed-off-by: Ana Dimova <a.dimova@prosyst.bg>

### CODE REVIEW

File: bundles/automation/org.eclipse.smarthome.automation.importers/.classpath [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.7"/>
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
+	<classpathentry kind="src" path="src/main/java"/>	<classpathentry kind="output" path="target/classes"/>
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.importers/.classpath]: there is a newline missing

**Comment by kaikreuzer** [comment]
> May be the name of the module can be fileproviders instead of importers

I agree that importers is misleading as the issue is only about providers. How about changing the bundle name to org.eclipse.smarthome.automation.provider.file?

Looking at the current bundles, I think we should also rename org.eclipse.smarthome.automation.providers to org.eclipse.smarthome.automation.provider and org.eclipse.smarthome.automation.commands to org.eclipse.smarthome.automation.command, what do you think?

File: bundles/automation/org.eclipse.smarthome.automation.importers/META-INF/MANIFEST.MF [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+Bundle-Name: Eclipse SmartHome Automation Importers
+Import-Package: 
+ org.apache.commons.lang,
+ org.eclipse.smarthome.automation,
+ org.eclipse.smarthome.automation.parser,
+ org.eclipse.smarthome.automation.template,
+ org.eclipse.smarthome.automation.type,
+ org.eclipse.smarthome.config.core,
+ org.eclipse.smarthome.core.common.registry,
+ org.eclipse.smarthome.io.console,
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.importers/META-INF/MANIFEST.MF]: what are the console imports good for?

File: bundles/automation/org.eclipse.smarthome.automation.importers/src/main/java/org/eclipse/smarthome/automation/internal/importers/AbstractImporter.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+import org.eclipse.smarthome.automation.type.ModuleTypeProvider;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceReference;
+import org.osgi.util.tracker.ServiceTracker;
+import org.osgi.util.tracker.ServiceTrackerCustomizer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is base for {@link RuleProvider}, {@link ModuleTypeProvider} and {@link TemplateProvider} which are
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.importers/src/main/java/org/eclipse/smarthome/automation/internal/importers/AbstractImporter.java]: Issue #1569 only speaks about module types and templates, but not about rules. Note that we have import mechanisms for rules already in place, which add them to the internal database. I would therefore reduce the scope here to only handle module types and templates.

File: bundles/automation/org.eclipse.smarthome.automation.importers/src/main/java/org/eclipse/smarthome/automation/internal/importers/AbstractImporter.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+ * responsible for importing the automation objects from local file system.
+ * <p>
+ * It provides functionality for tracking {@link Parser} services by implementing {@link ServiceTrackerCustomizer} and
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.importers/src/main/java/org/eclipse/smarthome/automation/internal/importers/AbstractImporter.java]: In the light of https://github.com/eclipse/smarthome/issues/1568, shouldn't you use proper DS right away?

File: bundles/automation/org.eclipse.smarthome.automation.importers/src/main/java/org/eclipse/smarthome/automation/internal/importers/Activator.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    @Override
+    public void start(BundleContext bc) throws Exception {
+        this.bc = bc;
+        try {
+            Filter filter = bc.createFilter("(|(objectClass=" + TemplateRegistry.class.getName() + ")(objectClass="
+                    + ModuleTypeRegistry.class.getName() + "))");
+            tracker = new ServiceTracker(bc, filter, this);
+            tracker.open();
+        } catch (InvalidSyntaxException e) {
+            e.printStackTrace();
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.importers/src/main/java/org/eclipse/smarthome/automation/internal/importers/Activator.java]: no

File: bundles/automation/org.eclipse.smarthome.automation.importers/src/main/java/org/eclipse/smarthome/automation/internal/importers/Activator.java [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+    /**
+     * This method initialize importers for importing automation objects from local file system.
+     */
+    @SuppressWarnings("unchecked")
+    @Override
+    public void start(BundleContext bc) throws Exception {
+        this.bc = bc;
+        try {
+            Filter filter = bc.createFilter("(|(objectClass=" + TemplateRegistry.class.getName() + ")(objectClass="
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.importers/src/main/java/org/eclipse/smarthome/automation/internal/importers/Activator.java]: This bundle must not depend on either the template nor the module type registry. It should merely register providers as a service.

**Comment by kaikreuzer** [comment]
This PR looks overly complicated to me - reducing the scope to module types and templates and using proper DS will imho heavily reduce the lines of code.

What's more, I would expect the implementation to actively watch the folder for changes, using AbstractWatchService.

**Comment by danchom** [comment]
I'm agree about removing 's' from package names.

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/ESH-INF/config/automationmoduletyperesources.xml [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+<config-description:config-descriptions xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:config-description="http://eclipse.org/smarthome/schemas/config-description/v1.0.0"
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/config-description/v1.0.0
+        http://eclipse.org/smarthome/schemas/config-description-1.0.0.xsd">
+
+	<config-description uri="system:automationresources:moduletypes">
+		<parameter name="roots" type="text" required="true">
+			<label>Automation Resource File Directories</label>
+			<description><![CDATA[
+                 <p>A string with relative paths to the local directories that containe automation resource files, comma separated.</p>
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/ESH-INF/config/automationmoduletyperesources.xml]: containe => contains

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/ESH-INF/config/automationtemplateresources.xml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+<config-description:config-descriptions xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xmlns:config-description="http://eclipse.org/smarthome/schemas/config-description/v1.0.0"
+    xsi:schemaLocation="http://eclipse.org/smarthome/schemas/config-description/v1.0.0
+        http://eclipse.org/smarthome/schemas/config-description-1.0.0.xsd">
+
+    <config-description uri="system:automationresources:templates">
+        <parameter name="roots" type="text" required="true">
+            <label>Automation Resource File Directories</label>
+            <description><![CDATA[
+                 <p>A string with relative paths to the local directories that containe automation resource files, comma separated.</p>
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/ESH-INF/config/automationtemplateresources.xml]: containe => contains

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/ESH-INF/config/automationtemplateresources.xml [hunk 9]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<config-description:config-descriptions xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xmlns:config-description="http://eclipse.org/smarthome/schemas/config-description/v1.0.0"
+    xsi:schemaLocation="http://eclipse.org/smarthome/schemas/config-description/v1.0.0
+        http://eclipse.org/smarthome/schemas/config-description-1.0.0.xsd">
+
+    <config-description uri="system:automationresources:templates">
+        <parameter name="roots" type="text" required="true">
+            <label>Automation Resource File Directories</label>
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/ESH-INF/config/automationtemplateresources.xml]: I don't think that the label should be equal for the both services.

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/OSGI-INF/TemplateFileProvider.xml [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+-->
+<scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0" immediate="true"
+    name="org.eclipse.smarthome.automation.internal.provider.file.TemplateFileProvider">
+    <implementation class="org.eclipse.smarthome.automation.internal.provider.file.MultipleTemplateFileProvider" />
+    <reference bind="addParser" cardinality="0..n" interface="org.eclipse.smarthome.automation.parser.Parser" name="Parser" policy="static" unbind="removeParser" target="(parser.type=parser.template)"/>
+    <service>
+        <provide interface="org.eclipse.smarthome.automation.template.TemplateProvider" />
+    </service>
+    <property name="service.config.description.uri" type="String" value="system:automation:resources"/>
+    <property name="service.config.label" type="String" value="Automation Resource File Directories"/>
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/OSGI-INF/TemplateFileProvider.xml]: See above (same label)

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+                for (File f : files) {
+                    importResources(f);
+                }
+            } else {
+                try {
+                    URL url = file.toURI().toURL();
+                    String parserType = getParserType(url);
+                    importFile(parserType, url);
+                } catch (MalformedURLException e) {
+                    // should not happen
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: If it should not happen, we can write a log message if it happen and not silently ignore the exception.

File: bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+    }
+
+    protected void remove(File path) {
+        try {
+            URL url = path.toURI().toURL();
+            synchronized (providerPortfolio) {
+                removeElements(providerPortfolio.remove(url));
+            }
+        } catch (MalformedURLException e) {
+            // should not happen
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java]: If it should not happen, we can write a log message if it happen and not silently ignore the exception.

File: bundles/automation/org.eclipse.smarthome.automation.sample.extension.json/SmartHome Runtime Sample Rules.launch [hunk 13]
```diff
@@ -1,34 +0,0 @@
-<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.sample.extension.json/SmartHome Runtime Sample Rules.launch]: Do you need to touch that file at all?

Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.sample.extension.json/SmartHome Runtime Sample Rules.launch]: I haven't touched this file

Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.sample.extension.json/SmartHome Runtime Sample Rules.launch]: Then the git history lies ;)

`git show --stat --format=raw e1701808da980cd24c436417c032c9679132464a`

```
commit e1701808da980cd24c436417c032c9679132464a
tree d942310d1d2441a28d6b3a86e5158c40b0e90cd6
parent 93b00d106b5839feb6143b532e7c003590957e36
author Ana Dimova <a.dimova@prosyst.bg> 1467962736 +0300
committer Ana Dimova <a.dimova@prosyst.bg> 1467962736 +0300

    Refactor #1569 implemetation
    
    Signed-off-by: Ana Dimova <a.dimova@prosyst.bg>

 bundles/automation/.project                                                                                                                                         |  17 +++++++++
 bundles/automation/.settings/org.eclipse.core.resources.prefs                                                                                                       |   2 ++
 bundles/automation/.settings/org.eclipse.m2e.core.prefs                                                                                                             |   4 +++
 bundles/automation/org.eclipse.smarthome.automation.importers/src/main/java/org/eclipse/smarthome/automation/internal/importers/AbstractImporter.java               | 124 ---------------------------------------------------------------
 bundles/automation/org.eclipse.smarthome.automation.importers/src/main/java/org/eclipse/smarthome/automation/internal/importers/Activator.java                      | 164 -----------------------------------------------------------------------------------
 bundles/automation/org.eclipse.smarthome.automation.importers/src/main/java/org/eclipse/smarthome/automation/internal/importers/ModuleTypeImporter.java             | 215 ------------------------------------------------------------------------------------------------------------
 bundles/automation/org.eclipse.smarthome.automation.importers/src/main/java/org/eclipse/smarthome/automation/internal/importers/TemplateImporter.java               | 208 ---------------------------------------------------------------------------------------------------------
 bundles/automation/{org.eclipse.smarthome.automation.importers => org.eclipse.smarthome.automation.provider.file}/.classpath                                        |   3 +-
 bundles/automation/{org.eclipse.smarthome.automation.importers => org.eclipse.smarthome.automation.provider.file}/.project                                          |   2 +-
 bundles/automation/{org.eclipse.smarthome.automation.importers => org.eclipse.smarthome.automation.provider.file}/META-INF/MANIFEST.MF                              |  10 +++---
 bundles/automation/org.eclipse.smarthome.automation.provider.file/OSGI-INF/ModuleTypeFileProvider.xml                                                               |  18 ++++++++++
 bundles/automation/org.eclipse.smarthome.automation.provider.file/OSGI-INF/TemplateFileProvider.xml                                                                 |  18 ++++++++++
 bundles/automation/{org.eclipse.smarthome.automation.importers => org.eclipse.smarthome.automation.provider.file}/about.html                                        |   0
 bundles/automation/{org.eclipse.smarthome.automation.importers => org.eclipse.smarthome.automation.provider.file}/build.properties                                  |   0
 bundles/automation/{org.eclipse.smarthome.automation.importers => org.eclipse.smarthome.automation.provider.file}/pom.xml                                           |   0
 bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java   | 289 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/ModuleTypeFileProvider.java |  92 +++++++++++++++++++++++++++++++++++++++++++++++
 bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/TemplateFileProvider.java   |  91 ++++++++++++++++++++++++++++++++++++++++++++++
 bundles/automation/org.eclipse.smarthome.automation.sample.extension.json/SmartHome Runtime Sample Rules.launch                                                     |  34 ------------------
 19 files changed, 539 insertions(+), 752 deletions(-)
```

**Comment by adimova** [comment]
replaced with new clean PR https://github.com/eclipse/smarthome/pull/2070

**Pull request closed without merging**